keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,425 @@
+
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia-k80
+spec:
+  labelSelector:
+    - matchExpressions:
+        - key: "Type"
+          operator: "Eq"
+          values:
+            - "nvidia-tesla-k80"
+
+kind: ResourceClass
+metadata:
+  name: nvidia-p100
+spec:
+  labelSelector:
+    - matchExpressions:
+        - key: "Type"
+          operator: "Eq"
+          values:
+            - "nvidia-tesla-p100"
+```
+
+Later on, suppose the cluster admins add a new GPU node group with a new version of GPU device plugin that exports another resource property "Nvlink" which will be set true for nvidia-tesla-p100 GPUs connected through nvlinks. To utilize this new feature, the cluster admins define the following new ResourceClass with nvlink constraints:
+
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia-p100-nvlink
+spec:
+  labelSelector:
+    - matchExpressions:
+        - key: "Type"
+          operator: "Eq"
+          values:
+            - "nvidia-tesla-p100"
+        - key: "Nvlink"
+          operator: "Eq"
+          values:
+            - "true"
+```
+
+Now we face the question that whether the scheduler should allow Pods requesting
+"nvidia-p100" to land on a node in this new GPU node groups. So far, we have
+received different feedbacks on this question. In some use cases, users would
+like to have minimum matching behavior that as long as the underlying hardware
+matches the minimal requirements specified through ResourceClass contraints,
+they want to allow Pods to be scheduled on the hardware. On the other hand, some users desire to reserve expensive hardware resources for users who explicitly request them.
+We feel both use cases are valid requirements. Allowing a ComputeResource to match
+multiple ResourceClasses as long as it matches their matching constraints
+perhaps yields least surprising behavior to users and also simplies upgrade
+scenario as new resource properties are introduced into the system. Therefore we
+support this behavior by default. To also provide an easy way for cluster admins
+to reserve expensive compute resources and control their access with resource
+quota, we propose to include a Priority field in ResourceClass API.
+By default, the value of this field is set to zero, but cluster admins can set
+it to a higher value, which would prevent its matching compute resources from
+being matched by lower priority ResourceClasses. i.e.,
+when a ComputeResource matches multiple ResourceClasses with different Priority values, the scheduler will choose those with the highest Priority.
+Supporting multiple ResourceClass matching also makes it easy to ensure that existing pods requesting resources through raw resource name can continue to be scheduled properly when administrators add ResourceClass in a cluster. To guarantee this, the scheduler may just consider raw resource as a special ResourceClass with empty resource metadata constraints.
```

Comment:
> scheduler may just consider raw resource as a special ResourceClass

This seems to conflict with the priority semantics above; if there's  a higher priority resource class that includes a raw resource, then raw-resource scheduling would be effectively disabled. Is that correct?

Reply:
will fix the wording. Whether it is a raw resource or resource class request, scheduler will figure this out using node info in the cache and irrespective of resource class priority, if raw resource is specified, scheduling will be done.